lib/filterx/filterx-object.h

```diff
@@ -483,4 +483,64 @@ filterx_object_is_type_or_ref(FilterXObject *object, FilterXType *type)
   return _filterx_object_is_type(object, type);
 }
 
+/*
+ * Make sure mutable objects are encapsulated into a FilterXRef.  To be
+ * called as the first thing before an object can be assigned to a variable
+ * or stored in a dict/list.
+ *
+ * NOTE: potentially replaces *value with a FilterXRef, sinking the passed
+ * reference, returning a FilterXRef instead.
+ */
+static inline void
+filterx_object_cow_wrap(FilterXObject **o)
+{
+  FilterXObject *value = *o;
+  if (!value || value->readonly || !_filterx_type_is_referenceable(value->type))
+    return;
+  *o = _filterx_ref_new(value);
+}
+
+
+/* NOTE: potentially replaces *value with a FilterXRef, sinking the passed
+ * reference.  It replaces the copied object as a return value */
+static inline FilterXObject *
+filterx_object_cow_fork(FilterXObject **o)
```

Comment:
As we are planning to implement dirty tracking on top of FilterXRef, I think this API is no longer only about CoW.

IMHO `mutable`, `mut`, `ref`, or `reference` is the name we are looking for here, copy-on-write is only just one of the features it provides.

Reply:
I didn't change the naming here, but moved the set_parent_container/unset_parent functions back to filterx_ref namespace.